File: pdui-express/lib/core/PDUINavigation.ts
import { PDUIFn } from "../models/PDUIFn";
import type { PBFnExpression, PBNavigation } from "../proto-out/pdui-proto-out";

export class PDUINavigation extends PDUIFn {
    expressionId: string;

    constructor({ expressionId }: { expressionId: string }) {
        super({
            identity: "navigation",
            name: undefined,
            params: {},
            body: {
                navigation: {
                    identity: "navigation",
                    payload: {
                        oneofKind: "navigation",
                        navigation: { expressionId: expressionId },
                    },
                },
            },
        });

        this.expressionId = expressionId;
    }

    override toPB(): PBNavigation {
        return { expressionId: this.expressionId };
    }

    override toFnExpression(): PBFnExpression {
        return {
            identity: "navigation",
            payload: {
                oneofKind: "navigation",
                navigation: { expressionId: this.expressionId },
            },
        };
    }
}

File: pdui_flutter/lib/core/pdui_navigation.dart
import 'package:flutter/material.dart';
import 'package:pdui_flutter/screens/pdui_screen.dart';

import '../models/pdui_fn.dart';
import '../proto_out/pdui-proto-out.pb.dart';
import 'pdui.dart';
import 'pdui_context.dart';

class PduiNavigation extends PduiFn {
  PBNavigation pbNavigation;

  PduiNavigation(this.pbNavigation) : super("navigation");

  @override
  resolve() {
    Navigator.of(PduiContext.navigatorKey.currentContext!).push(
      MaterialPageRoute(
        builder: (context) => PduiScreen(
          expressionId: pbNavigation.expressionId,
          useCache: Pdui.useCache,
        ),
      ),
    );
  }
}
